src/main/plugin/iso19139.ca.HNAP/schematron/schematron-rules-common.sch

```diff
@@ -519,6 +521,11 @@
 
       <sch:assert test="$mapWMSCount &lt;= 2">$loc/strings/MapResourcesWMSNumber</sch:assert>
       <sch:assert test="$mapWMSCount = 0 or $mapWMSCount = 2 or $mapWMSCount &gt; 2">$loc/strings/MapResourcesWMS</sch:assert>
+
+      <sch:let name="noneSpatialType" value="not(string(../../gmd:identificationInfo/gmd:MD_DataIdentification/gmd:spatialRepresentationType))"/>
+
+      <sch:assert test="not($noneSpatialType) or XslUtilHnap:hasNoDuplicatedOnlineResource($onlineResources)">$loc/strings/hasDuplicatedOnlineResource</sch:assert>
```

Comment:
I would expect that you should be able to identify the duplicates using xslt without needing to use java?

Here is some quick code created from google search and chat gpt. - But I have not tested it.

```
<xsl:function xmlns:sch="http://purl.oclc.org/dsdl/schematron" name="geonet:hasDuplicateResources" as="xs:boolean">
  <xsl:param name="elements" as="element()*"/>
  
  <xsl:key name="groupKey" match="yourElement" use="concat(gmd:url/gco:CharacterString/text(), '|', gmd:name/gco:CharacterString/text())" />

  <xsl:value-of select="boolean($elements[key('groupKey', concat(gmd:url/gco:CharacterString/text(), '|', gmd:name/gco:CharacterString/text())][2])"/>
</xsl:function>
```

```
<!-- Example usage -->
<sch:let name="hasDuplicates" select="geonet:hasDuplicateResources($onlineResources)"/>
```

Reply:
I just noticed - the unique resource is based on the url and language - not url and name so you may want to fix that as well.

Reply:
Looks like the key needs to be outside the function

The following should work

```
  <xsl:key name="groupKey"
           match="*//gmd:transferOptions/gmd:MD_DigitalTransferOptions/gmd:onLine/gmd:CI_OnlineResource[gmd:linkage/gmd:URL]"
           use="concat(gmd:linkage/gmd:URL/text(), '|', gmd:name/gco:CharacterString/text())"/>

  <xsl:function xmlns:sch="http://purl.oclc.org/dsdl/schematron"
                name="geonet:hasDuplicateResources"
                as="xs:boolean">
    <xsl:param name="elements" as="element()*"/>

    <xsl:variable name="duplicateResources">
      <xsl:for-each-group select="$elements"
                          group-by="concat(gmd:linkage/gmd:URL/text(), '|', gmd:name/gco:CharacterString/text())">

        <xsl:if test="count(current-group()) &gt; 1">
          <xsl:if test="exists(tokenize(current-group()[1]/gmd:description/gco:CharacterString/text(), ';')[3])">
            <xsl:value-of select="concat(tokenize(current-group()[1]/gmd:description/gco:CharacterString/text(), ';')[3], ';')"/>
          </xsl:if>
          <xsl:value-of select="current-group()[1]/gmd:linkage/gmd:URL/text()"/>
        </xsl:if>
      </xsl:for-each-group>
    </xsl:variable>

    <xsl:choose>
      <xsl:when test="exists($duplicateResources)">
        <xsl:message>
          <xsl:value-of select="$duplicateResources"/>
        </xsl:message>
        <xsl:sequence select="true()"/>
      </xsl:when>
      <xsl:otherwise>
        <xsl:sequence select="false()"/>
      </xsl:otherwise>
    </xsl:choose>
  </xsl:function>
```

Note: you may want to change the logic so that if a duplicate is identified then it provides the duplicate url in the error message so that it is easier for the user to know where the duplicate is located.

Reply:
I tested it. It works good. Thank you. I also made small adjust to have url+language as you suggested in the earlier comment. 

I have commit the latest code and removed all Java related checking logic

Reply:
I commit another version which will render language,url instead of just boolean.

![image](https://github.com/metadata101/iso19139.ca.HNAP/assets/74916635/ac751d27-4be4-4b3a-b7b5-cffe6e9b24ea)